### PR TITLE
Support remote config from extraction pipelines [DEGR-2723]

### PR DESCRIPTION
**Pull request by kbrattli**
Symmtry (TESTED AND WORKS): https://github.com/cognitedata/symmetry-connector-dotnet/pull/23
SimConnect (TESTED AND WORKS): https://github.com/cognitedata/simconnect-dotnet/pull/137
DWSIM (TESTED AND WORKS): https://github.com/cognitedata/dwsim-connector-dotnet/pull/34
PetroSim (TESTED AND WORKS): https://github.com/cognitedata/petrosim-connector-dotnet/pull/30

---------------------------------

https://cognitedata.atlassian.net/browse/DEGR-2723

**ConnectorBase.cs**
This allows us to check for if the Remote Config from fusion.cognite has been updated. Every connector has to be updated but it is small change, simply do this:
```
public class SymmetryConnector : ConnectorBase<FullConfig> //------- Pay attention here
{

    private readonly ModelLibrary _modelLibrary;

    private readonly ConfigurationLibrary _configurationLibrary;

    private readonly FullConfig _config;
    private readonly ILogger<SymmetryConnector> _logger;

    private readonly SymmetryClient _symmetry;

    private readonly SimulationRunner _simulationRunner;

    private readonly SimulationScheduler _simulationScheduler;

    private readonly ExtractionPipeline _pipeline;

    private readonly string _version;

    public SymmetryConnector(
         CogniteDestination cdf,
         FullConfig config,
        ModelLibrary modelLibrary,
        ConfigurationLibrary configurationLibrary,
        SymmetryClient symmetry,
        SimulationScheduler scheduler,
        SimulationRunner runner,
        ExtractionPipeline pipeline,
        ILogger<SymmetryConnector> logger,
        RemoteConfigManager<FullConfig> remoteConfigManager) //------ Pay attention here
         : base(cdf, config.Connector, new List<SimulatorConfig> { config.Simulator }, logger, remoteConfigManager) //--- Pay attention here
    {
```
Then add this in the run(): `taskList.Add(CheckForNewConfig(linkedToken));`. Important to check that every task uses linkedToken and not token in Run(), if one of them uses token, `CheckForNewConfig` will not work.

Also remember to change *ConnectorRuntime.cs* in every connector. Look here for guidance: https://github.com/cognitedata/symmetry-connector-dotnet/pull/23. The important idea is that the NewConfigException needs to propagate the whole way up to get catched and then can restart the connector. So if you add a new catch for general exception, it is important that you write: `catch (Exception e) when (e is not NewConfigDetected)`

----------------------------------------------------
**ExtractionPipeline.cs**
This helper function first extracts the local config, then it will retrieve the "minimum config" aka the "cognite and type" part of the cognite. Then if type is remote, addRemoteConfig will get the FullConfig from Monitor Extraction Pipelines. To do this it has to be linked to a Extraction Pipeline with for example:
```
type: remote
cognite:
  extraction-pipeline:
    pipeline-id: symmetry-extraction-pipeline-kenneth
```
if type is local, then it will just get the local config instead.

This function has to be used in every `ConnectorRuntime.cs` see below for example, also add the exception handling.

### CODE REVIEW

File: Cognite.Simulator.Utils/ConnectorBase.cs [hunk 1]
```diff
@@ -284,6 +289,30 @@ await UpdateIntegrationRows(false, token)
...
+        
+        /// <summary>
+        /// Task that runs in a loop, checking for new config in extraction pipelines
+        /// </summary>
+        public async Task CheckForNewConfig(CancellationToken token)
+        {
+            while (!token.IsCancellationRequested)
+            {
+                await Task
+                    .Delay(900000, token) // Run every 15 minutes
```
Comment (kbrattli) [inline Cognite.Simulator.Utils/ConnectorBase.cs:301]: use constant variable that is 9000

Comment (abdullah-cognite) [inline Cognite.Simulator.Utils/ConnectorBase.cs:301]: `const 15_MINUTES = 900000`

Comment (kbrattli) [reply Cognite.Simulator.Utils/ConnectorBase.cs:301]: check if you have specified u want local config, and this config runs, what will happen

Comment (kbrattli) [reply Cognite.Simulator.Utils/ConnectorBase.cs]: Ok its fine

File: Cognite.Simulator.Utils/ExtractionPipeline.cs [hunk 2]
```diff
@@ -237,5 +239,47 @@ public static void AddExtractionPipeline(this IServiceCollection services, Conne
             services.AddSingleton(config.PipelineNotification);
             services.AddScoped<ExtractionPipeline>();
         }
+        /// <summary>
+        /// Uses a minimum config that needs "cognite" and type: "specified",
+        /// then uses that to either fetch remote Config Extraction Pipeline, or just adds
+        /// the local one based on if "type" is remote or local.
```
Comment (polomani) [inline Cognite.Simulator.Utils/ExtractionPipeline.cs:245]: This is quite hard to read :)
Maybe provide actual yaml examples?

File: Cognite.Simulator.Utils/ExtractionPipeline.cs [hunk 3]
```diff
@@ -237,5 +239,47 @@ public static void AddExtractionPipeline(this IServiceCollection services, Conne
...
+        {
+            var localConfig = services.AddConfig<T>(path, version);
+            var remoteConfig = new RemoteConfig
+            {
+                Type = localConfig.Type,
+                Cognite = localConfig.Cognite
+            };
+            
+            return await services.AddRemoteConfig<T>(
+                logger: null,
```
Comment (polomani) [inline Cognite.Simulator.Utils/ExtractionPipeline.cs:272]: fix the warning with ConfigureAwait(...)

File: Cognite.Simulator.Utils/ConnectorBase.cs [hunk 4]
```diff
@@ -284,6 +291,33 @@ await UpdateIntegrationRows(false, token)
                     .ConfigureAwait(false);
             }
         }
+        
+        /// <summary>
+        /// Task that runs in a loop, checking for new config in extraction pipelines
+        /// </summary>
+        public async Task CheckForNewConfig(CancellationToken token)
```
Comment (polomani) [inline Cognite.Simulator.Utils/ConnectorBase.cs:298]:
```suggestion
        public async Task CheckRemoteConfig(CancellationToken token)
```
to make it clear it's remote

File: Cognite.Simulator.Utils/ConnectorBase.cs [hunk 5]
```diff
@@ -284,6 +291,33 @@ await UpdateIntegrationRows(false, token)
...
+        {
+            while (!token.IsCancellationRequested)
+            {
+                await Task
+                    .Delay(FIFTEEN_MIN, token) // Run every 15 minutes
+                    .ConfigureAwait(false);
+                _logger.LogDebug("Checking remote config");
```
Comment (polomani) [inline Cognite.Simulator.Utils/ConnectorBase.cs:305]:
```suggestion
                _logger.LogDebug("Checking remote config updates");
```

File: Cognite.Simulator.Utils/ExtractionPipeline.cs [hunk 6]
```diff
@@ -240,9 +240,16 @@ public static void AddExtractionPipeline(this IServiceCollection services, Conne
             services.AddScoped<ExtractionPipeline>();
         }
         /// <summary>
-        /// Uses a minimum config that needs "cognite" and type: "specified",
-        /// then uses that to either fetch remote Config Extraction Pipeline, or just adds
-        /// the local one based on if "type" is remote or local.
+        /// if type: remote is instead of local, it will fetch config from Fusion, if local, it will use the local config file
+        /// Example from config.yml if one wishes to use the remote config from Fusion:
```
Comment (polomani) [inline Cognite.Simulator.Utils/ExtractionPipeline.cs:244]:
```suggestion
        /// Example from config.yml using the remote config from Fusion
```

File: version [hunk 7]
```diff
@@ -1 +1 @@
-1.0.0-alpha-013
+1.0.0-alpha-014
```
Comment (sagar-thalwar) [inline version:1]:
```suggestion
2.0.0-alpha-001
```

**Review by abdullah-cognite — APPROVED** [review]

**Pull request merged**